File: runinator-command-center/src/utils/wdl-expression.ts
import type { JsonRecord } from "../types/models";

type TokenKind = "ident" | "string" | "number" | "op" | "punct" | "eof";

interface Token {
  kind: TokenKind;
  text: string;
}

const expressionKeys = ["$ref", "$concat", "$coalesce", "$literal", "$to_string", "$to_json_string", "$node"];

export function expressionJsonToWdl(value: unknown): string {
  if (value === null) return "null";
  if (typeof value === "boolean" || typeof value === "number") return String(value);
  if (typeof value === "string") return secretRefToWdl(value) ?? quote(value);
  if (Array.isArray(value)) return `[${value.map(expressionJsonToWdl).join(", ")}]`;
  if (!isRecord(value)) return "null";

  const keys = Object.keys(value);
  if (keys.length === 1) {
    if (isRecord(value.$ref)) return refToWdl(value.$ref);
    if (Array.isArray(value.$concat)) return joinOperator(value.$concat, " ++ ");
    if (Array.isArray(value.$coalesce)) return joinOperator(value.$coalesce, " ?? ");
    if (Object.prototype.hasOwnProperty.call(value, "$to_string")) return `string(${expressionJsonToWdl(value.$to_string)})`;
    if (Object.prototype.hasOwnProperty.call(value, "$to_json_string")) return `json(${expressionJsonToWdl(value.$to_json_string)})`;
  }

  const entries = Object.entries(value).map(([key, nested]) => `${objectKey(key)}: ${expressionJsonToWdl(nested)}`);
  return `{ ${entries.join(", ")} }`;
}

export function parseWdlExpression(source: string): unknown {
  const parser = new Parser(tokenize(source));
  const value = parser.parseExpression();
  parser.expect("eof");
  return value;
}

export function isWorkflowExpressionValue(value: unknown): value is JsonRecord {
  return isRecord(value) && expressionKeys.some((key) => Object.prototype.hasOwnProperty.call(value, key));
}

function joinOperator(items: unknown[], operator: string): string {
  return items.map((item) => wrapBinaryPart(expressionJsonToWdl(item))).join(operator);
}

function wrapBinaryPart(value: string): string {
  return /\s(?:\+\+|\?\?)\s/.test(value) ? `(${value})` : value;
}

function refToWdl(ref: JsonRecord): string {
  if (Array.isArray(ref.params)) return appendPath("params", ref.params);
  if (Array.isArray(ref.prev)) return appendPath("prev", ref.prev);
  if (Array.isArray(ref.workflow)) return appendPath("run", ref.workflow);
  if (Array.isArray(ref.config)) return appendPath("config", ref.config);
  if (typeof ref.node === "string" && Array.isArray(ref.output)) return appendPath(ref.node, ref.output);
  return objectLiteral(ref);
}

function appendPath(head: string, path: unknown[]): string {
  return [head, ...path.map((segment) => String(segment))].join(".");
}

function objectLiteral(record: JsonRecord): string {
  const entries = Object.entries(record).map(([key, value]) => `${objectKey(key)}: ${expressionJsonToWdl(value)}`);
  return `{ ${entries.join(", ")} }`;
}

function objectKey(key: string): string {
  return /^[A-Za-z_][A-Za-z0-9_]*$/.test(key) ? key : quote(key);
}

function quote(value: string): string {
  return JSON.stringify(value);
}

function secretRefToWdl(value: string): string | null {
  if (!value.startsWith("secret://")) return null;
  const rest = value.slice("secret://".length);
  const [scope, ...name] = rest.split("/");
  if (!scope || name.length === 0 || ![scope, ...name].every((part) => /^[A-Za-z_][A-Za-z0-9_]*$/.test(part))) return null;
  return ["secret", scope, ...name].join(".");
}

function tokenize(source: string): Token[] {
  const tokens: Token[] = [];
  let index = 0;
  while (index < source.length) {
    const char = source[index];
    if (/\s/.test(char)) {
      index += 1;
      continue;
    }
    if (source.startsWith("++", index) || source.startsWith("??", index)) {
      tokens.push({ kind: "op", text: source.slice(index, index + 2) });
      index += 2;
      continue;
    }
    if ("{}[]():,.".includes(char)) {
      tokens.push({ kind: "punct", text: char });
      index += 1;
      continue;
    }
    if (char === "\"") {
      const start = index;
      index += 1;
      let escaped = false;
      while (index < source.length) {
        const next = source[index++];
        if (escaped) {
          escaped = false;
          continue;
        }
        if (next === "\\") {
          escaped = true;
          continue;
        }
        if (next === "\"") break;
      }
      tokens.push({ kind: "string", text: source.slice(start, index) });
      continue;
    }
    const number = source.slice(index).match(/^-?\d+(?:\.\d+)?/);
    if (number) {
      tokens.push({ kind: "number", text: number[0] });
      index += number[0].length;
      continue;
    }
    const ident = source.slice(index).match(/^[A-Za-z_$][A-Za-z0-9_$-]*/);
    if (ident) {
      tokens.push({ kind: "ident", text: ident[0] });
      index += ident[0].length;
      continue;
    }
    throw new Error(`Unexpected character ${char}`);
  }
  tokens.push({ kind: "eof", text: "" });
  return tokens;
}

class Parser {
  private index = 0;

  constructor(private readonly tokens: Token[]) {}

  parseExpression(): unknown {
    return this.parseCoalesce();
  }

  expect(kind: TokenKind, text?: string): Token {
    const token = this.peek();
    if (token.kind !== kind || (text !== undefined && token.text !== text)) {
      throw new Error(text ? `Expected ${text}` : `Expected ${kind}`);
    }
    this.index += 1;
    return token;
  }

  private parseCoalesce(): unknown {
    const parts = [this.parseConcat()];
    while (this.match("op", "??")) parts.push(this.parseConcat());
    return parts.length === 1 ? parts[0] : { "$coalesce": parts };
  }

  private parseConcat(): unknown {
    const parts = [this.parsePrimary()];
    while (this.match("op", "++")) parts.push(this.parsePrimary());
    return parts.length === 1 ? parts[0] : { "$concat": parts };
  }

  private parsePrimary(): unknown {
    const token = this.peek();
    if (this.match("punct", "(")) {
      const value = this.parseExpression();
      this.expect("punct", ")");
      return value;
    }
    if (this.match("punct", "{")) return this.parseObject();
    if (this.match("punct", "[")) return this.parseArray();
    if (token.kind === "string") {
      this.index += 1;
      return JSON.parse(token.text);
    }
    if (token.kind === "number") {
      this.index += 1;
      return token.text.includes(".") ? Number.parseFloat(token.text) : Number.parseInt(token.text, 10);
    }
    if (token.kind === "ident") return this.parseIdentPrimary();
    throw new Error("Expected expression");
  }

  private parseIdentPrimary(): unknown {
    const head = this.expect("ident").text;
    if (head === "true") return true;
    if (head === "false") return false;
    if (head === "null") return null;
    if ((head === "string" || head === "json") && this.match("punct", "(")) {
      const nested = this.parseExpression();
      this.expect("punct", ")");
      return head === "string" ? { "$to_string": nested } : { "$to_json_string": nested };
    }
    const path = [head];
    while (this.match("punct", ".")) path.push(this.expectPathSegment());
    return lowerPath(path);
  }

  private parseObject(): JsonRecord {
    const record: JsonRecord = {};
    while (!this.match("punct", "}")) {
      const keyToken = this.peek();
      let key: string;
      if (keyToken.kind === "string") {
        key = JSON.parse(this.expect("string").text);
      } else {
        key = this.expect("ident").text;
      }
      let value: unknown;
      if (this.match("punct", ":")) {
        value = this.parseExpression();
      } else {
        value = lowerPath([key]);
      }
      record[key] = value;
      if (this.match("punct", ",")) continue;
      this.expect("punct", "}");
      break;
    }
    return record;
  }

  private parseArray(): unknown[] {
    const items: unknown[] = [];
    while (!this.match("punct", "]")) {
      items.push(this.parseExpression());
      if (this.match("punct", ",")) continue;
      this.expect("punct", "]");
      break;
    }
    return items;
  }

  private expectPathSegment(): string {
    const token = this.peek();
    if (token.kind !== "ident" && token.kind !== "number") throw new Error("Expected path segment");
    this.index += 1;
    return token.text;
  }

  private match(kind: TokenKind, text?: string): boolean {
    const token = this.peek();
    if (token.kind !== kind || (text !== undefined && token.text !== text)) return false;
    this.index += 1;
    return true;
  }

  private peek(): Token {
    return this.tokens[this.index] ?? { kind: "eof", text: "" };
  }
}

function lowerPath(path: string[]): unknown {
  const [head, ...rest] = path;
  if (head === "params" || head === "prev" || head === "config") return { "$ref": { [head]: pathSegments(rest) } };
  if (head === "run" || head === "workflow") return { "$ref": { workflow: pathSegments(rest) } };
  if (head === "secret" && rest.length >= 2) return `secret://${rest[0]}/${rest.slice(1).join("/")}`;
  return { "$ref": { node: head, output: pathSegments(rest) } };
}

function pathSegments(path: string[]): Array<string | number> {
  return path.map((segment) => (/^\d+$/.test(segment) ? Number.parseInt(segment, 10) : segment));
}

function isRecord(value: unknown): value is JsonRecord {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: runinator-command-center-flutter/lib/core/utils/wdl_expression.dart
// port of core/utils/wdl-expression.ts.

import 'dart:convert';

import '../domain/json.dart';

enum _TokenKind { ident, string, number, op, punct, eof }

class _Token {
  const _Token(this.kind, this.text);

  final _TokenKind kind;
  final String text;
}

const List<String> _expressionKeys = [
  r'$ref',
  r'$concat',
  r'$coalesce',
  r'$literal',
  r'$to_string',
  r'$to_json_string',
  r'$node',
];

String expressionJsonToWdl(Object? value) {
  if (value == null) {
    return 'null';
  }

  if (value is bool || value is num) {
    return value.toString();
  }

  if (value is String) {
    return _secretRefToWdl(value) ?? _quote(value);
  }

  if (value is List) {
    return '[${value.map(expressionJsonToWdl).join(", ")}]';
  }

  if (!_isRecord(value)) {
    return 'null';
  }

  final record = value as JsonRecord;
  final keys = record.keys.toList();

  if (keys.length == 1) {
    final ref = record[r'$ref'];
    if (_isRecord(ref)) {
      return _refToWdl(ref as JsonRecord);
    }

    final concat = record[r'$concat'];
    if (concat is List) {
      return _joinOperator(concat, ' ++ ');
    }

    final coalesce = record[r'$coalesce'];
    if (coalesce is List) {
      return _joinOperator(coalesce, ' ?? ');
    }

    if (record.containsKey(r'$to_string')) {
      return 'string(${expressionJsonToWdl(record[r'$to_string'])})';
    }

    if (record.containsKey(r'$to_json_string')) {
      return 'json(${expressionJsonToWdl(record[r'$to_json_string'])})';
    }
  }

  final entries = record.entries.map((entry) => '${_objectKey(entry.key)}: ${expressionJsonToWdl(entry.value)}');
  return '{ ${entries.join(", ")} }';
}

JsonValue parseWdlExpression(String source) {
  final parser = _Parser(_tokenize(source));
  final value = parser.parseExpression();
  parser.expect(_TokenKind.eof);
  return value;
}

bool isWorkflowExpressionValue(Object? value) =>
    _isRecord(value) && _expressionKeys.any((key) => (value as JsonRecord).containsKey(key));

String _joinOperator(List<Object?> items, String operator) =>
    items.map((item) => _wrapBinaryPart(expressionJsonToWdl(item))).join(operator);

final RegExp _binaryOpPattern = RegExp(r'\s(?:\+\+|\?\?)\s');

String _wrapBinaryPart(String value) => _binaryOpPattern.hasMatch(value) ? '($value)' : value;

String _refToWdl(JsonRecord ref) {
  if (ref['params'] is List) {
    return _appendPath('params', ref['params'] as List);
  }

  if (ref['prev'] is List) {
    return _appendPath('prev', ref['prev'] as List);
  }

  if (ref['workflow'] is List) {
    return _appendPath('run', ref['workflow'] as List);
  }

  if (ref['config'] is List) {
    return _appendPath('config', ref['config'] as List);
  }

  if (ref['node'] is String && ref['output'] is List) {
    return _appendPath(ref['node'] as String, ref['output'] as List);
  }

  return _objectLiteral(ref);
}

String _appendPath(String head, List<Object?> path) =>
    [head, ...path.map((segment) => segment.toString())].join('.');

String _objectLiteral(JsonRecord record) {
  final entries = record.entries.map((entry) => '${_objectKey(entry.key)}: ${expressionJsonToWdl(entry.value)}');
  return '{ ${entries.join(", ")} }';
}

final RegExp _identPattern = RegExp(r'^[A-Za-z_][A-Za-z0-9_]*$');

String _objectKey(String key) => _identPattern.hasMatch(key) ? key : _quote(key);

String _quote(String value) => jsonEncode(value);

String? _secretRefToWdl(String value) {
  if (!value.startsWith('secret://')) {
    return null;
  }

  final rest = value.substring('secret://'.length);
  final parts = rest.split('/');

  if (parts.isEmpty) {
    return null;
  }

  final scope = parts.first;
  final name = parts.skip(1).toList();

  if (scope.isEmpty || name.isEmpty || ![scope, ...name].every((part) => _identPattern.hasMatch(part))) {
    return null;
  }

  return ['secret', scope, ...name].join('.');
}

List<_Token> _tokenize(String source) {
  final tokens = <_Token>[];
  var index = 0;

  while (index < source.length) {
    final char = source[index];

    if (RegExp(r'\s').hasMatch(char)) {
      index += 1;
      continue;
    }

    if (source.startsWith('++', index) || source.startsWith('??', index)) {
      tokens.add(_Token(_TokenKind.op, source.substring(index, index + 2)));
      index += 2;
      continue;
    }

    if ('{}[]():,.'.contains(char)) {
      tokens.add(_Token(_TokenKind.punct, char));
      index += 1;
      continue;
    }

    if (char == '"') {
      final start = index;
      index += 1;
      var escaped = false;

      while (index < source.length) {
        final next = source[index];
        index += 1;

        if (escaped) {
          escaped = false;
          continue;
        }

        if (next == '\\') {
          escaped = true;
          continue;
        }

        if (next == '"') {
          break;
        }
      }

      tokens.add(_Token(_TokenKind.string, source.substring(start, index)));
      continue;
    }

    final tail = source.substring(index);
    final numberMatch = RegExp(r'^-?\d+(?:\.\d+)?').matchAsPrefix(tail);

    if (numberMatch != null) {
      tokens.add(_Token(_TokenKind.number, numberMatch.group(0)!));
      index += numberMatch.group(0)!.length;
      continue;
    }

    final identMatch = RegExp(r'^[A-Za-z_$][A-Za-z0-9_$-]*').matchAsPrefix(tail);

    if (identMatch != null) {
      tokens.add(_Token(_TokenKind.ident, identMatch.group(0)!));
      index += identMatch.group(0)!.length;
      continue;
    }

    throw FormatException('Unexpected character $char');
  }

  tokens.add(const _Token(_TokenKind.eof, ''));
  return tokens;
}

class _Parser {
  _Parser(this._tokens);

  final List<_Token> _tokens;
  int _index = 0;

  JsonValue parseExpression() => _parseCoalesce();

  _Token expect(_TokenKind kind, [String? text]) {
    final token = _peek();

    if (token.kind != kind || (text != null && token.text != text)) {
      throw FormatException(text != null ? 'Expected $text' : 'Expected $kind');
    }

    _index += 1;
    return token;
  }

  JsonValue _parseCoalesce() {
    final parts = [_parseConcat()];

    while (_match(_TokenKind.op, '??')) {
      parts.add(_parseConcat());
    }

    return parts.length == 1 ? parts.first : {r'$coalesce': parts};
  }

  JsonValue _parseConcat() {
    final parts = [_parsePrimary()];

    while (_match(_TokenKind.op, '++')) {
      parts.add(_parsePrimary());
    }

    return parts.length == 1 ? parts.first : {r'$concat': parts};
  }

  JsonValue _parsePrimary() {
    final token = _peek();

    if (_match(_TokenKind.punct, '(')) {
      final value = parseExpression();
      expect(_TokenKind.punct, ')');
      return value;
    }

    if (_match(_TokenKind.punct, '{')) {
      return _parseObject();
    }

    if (_match(_TokenKind.punct, '[')) {
      return _parseArray();
    }

    if (token.kind == _TokenKind.string) {
      _index += 1;
      return asJsonValue(jsonDecode(token.text));
    }

    if (token.kind == _TokenKind.number) {
      _index += 1;
      return token.text.contains('.') ? double.parse(token.text) : int.parse(token.text);
    }

    if (token.kind == _TokenKind.ident) {
      return _parseIdentPrimary();
    }

    throw const FormatException('Expected expression');
  }

  JsonValue _parseIdentPrimary() {
    final head = expect(_TokenKind.ident).text;

    if (head == 'true') {
      return true;
    }

    if (head == 'false') {
      return false;
    }

    if (head == 'null') {
      return null;
    }

    if ((head == 'string' || head == 'json') && _match(_TokenKind.punct, '(')) {
      final nested = parseExpression();
      expect(_TokenKind.punct, ')');
      return head == 'string' ? {r'$to_string': nested} : {r'$to_json_string': nested};
    }

    final path = [head];

    while (_match(_TokenKind.punct, '.')) {
      path.add(_expectPathSegment());
    }

    return _lowerPath(path);
  }

  JsonValue _parseObject() {
    final record = <String, Object?>{};

    while (!_match(_TokenKind.punct, '}')) {
      final keyToken = _peek();
      String key;

      if (keyToken.kind == _TokenKind.string) {
        key = jsonDecode(expect(_TokenKind.string).text) as String;
      } else {
        key = expect(_TokenKind.ident).text;
      }

      JsonValue value;

      if (_match(_TokenKind.punct, ':')) {
        value = parseExpression();
      } else {
        value = _lowerPath([key]);
      }

      record[key] = value;

      if (_match(_TokenKind.punct, ',')) {
        continue;
      }

      expect(_TokenKind.punct, '}');
      break;
    }

    return asJsonValue(record);
  }

  JsonValue _parseArray() {
    final items = <Object?>[];

    while (!_match(_TokenKind.punct, ']')) {
      items.add(parseExpression());

      if (_match(_TokenKind.punct, ',')) {
        continue;
      }

      expect(_TokenKind.punct, ']');
      break;
    }

    return items;
  }

  String _expectPathSegment() {
    final token = _peek();

    if (token.kind != _TokenKind.ident && token.kind != _TokenKind.number) {
      throw const FormatException('Expected path segment');
    }

    _index += 1;
    return token.text;
  }

  bool _match(_TokenKind kind, [String? text]) {
    final token = _peek();

    if (token.kind != kind || (text != null && token.text != text)) {
      return false;
    }

    _index += 1;
    return true;
  }

  _Token _peek() => _index < _tokens.length ? _tokens[_index] : const _Token(_TokenKind.eof, '');
}

JsonValue _lowerPath(List<String> path) {
  final head = path.first;
  final rest = path.skip(1).toList();

  if (head == 'params' || head == 'prev' || head == 'config') {
    return {
      r'$ref': {head: _pathSegments(rest)},
    };
  }

  if (head == 'run' || head == 'workflow') {
    return {
      r'$ref': {'workflow': _pathSegments(rest)},
    };
  }

  if (head == 'secret' && rest.length >= 2) {
    return 'secret://${rest[0]}/${rest.skip(1).join("/")}';
  }

  return {
    r'$ref': {'node': head, 'output': _pathSegments(rest)},
  };
}

final RegExp _digitsOnly = RegExp(r'^\d+$');

List<Object> _pathSegments(List<String> path) =>
    path.map<Object>((segment) => _digitsOnly.hasMatch(segment) ? int.parse(segment) : segment).toList();

bool _isRecord(Object? value) => value is Map<String, Object?>;
